### PR TITLE
Add an internal UNet implementation

### PR DESCRIPTION
**Pull request by ianstenbit**
The modular UNet API that I had previously designed doesn't support the exact use case we have for our 3D OD model, so I started from scratch with an ultra-modular (and potentially very useless because it's so generic?) UNet design.

The default implementation of the blocks matches what we want for a starter CenterPillar model

### CODE REVIEW

File: keras_cv/models/__internal__/unet.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+        down_blocks: a list of (filter_count, num_blocks) tuples indicating the
+            number of filters and sub-blocks in each down block
+        up_blocks: a list of filter counts, one for each up block
+        down_block: a downsampling block
+        up_block: an upsampling block
+    """
+
+    def apply(x):
+        skip_connections = []
+        for filters, num_blocks in down_blocks:
```
Comment (tanzhenyu) [inline keras_cv/models/__internal__/unet.py:163]: maybe comment here what does `filters` and `num_blocks` represent

File: keras_cv/models/__internal__/unet.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+            x = Block(filters, downsample=False)(x)
+
+        return x
+
+    return apply
+
+
+def UpSampleBlock(filters):
+    def apply(x, lateral_input):
+        x = layers.Conv2DTranspose(
```
Comment (tanzhenyu) [inline keras_cv/models/__internal__/unet.py:114]: Is the UNET paper using UpSampling2D or Conv2DTranspose? Can this be configurable?

File: keras_cv/models/__internal__/unet.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+            filters,
+            3,
+            2,
+            padding="same",
+            kernel_initializer=initializers.VarianceScaling(),
+            kernel_regularizer=regularizers.L2(l2=1e-4),
```
Comment (tanzhenyu) [inline keras_cv/models/__internal__/unet.py:120]: I think in this codebase we usually don't put regularization during modeling part, but instead in training part. This makes modeling more generic and users can tune those hyperparameters without touch model parameters.

File: keras_cv/models/__internal__/unet.py [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        )(x)
+        x = layers.BatchNormalization(
+            synchronized=True,
```
Comment (tanzhenyu) [inline keras_cv/models/__internal__/unet.py:123]: Maybe we should ask whether we need to make this configurable or not. @Guowang @lengzq

File: keras_cv/models/__internal__/unet.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from tensorflow.keras import initializers
+from tensorflow.keras import layers
+from tensorflow.keras import regularizers
+
+
+def Block(filters, downsample):
```
Comment (tanzhenyu) [inline keras_cv/models/__internal__/unet.py:20]: Consider adding docstring (even though it's not public API)

**Review by tanzhenyu — COMMENTED** [review]
Thanks for the PR!

File: keras_cv/models/__internal__/unet.py (hunk 4 above)
Comment (lengzq) [reply keras_cv/models/__internal__/unet.py:123]: You meant synchronized batchnorm? Sure.

File: keras_cv/models/__internal__/unet.py (hunk 2 above)
Comment (lengzq) [reply keras_cv/models/__internal__/unet.py:114]: Bilinear is slower than ConvTranspose on TPU. We could support bilinear up sampling once we verify the unet performance.

Comment (ianstenbit) [reply keras_cv/models/__internal__/unet.py:114]: The idea here is that this UpSampleBlock (as well as the DownSampleBlock) are specific to this UNet.

When this is a public API, we could example call `UNet(...., down_block=MyCustomDownBlock)`
In that future, these particular up/down blocks would be available as e.g. a CenterPillarDownBlock

File: keras_cv/models/__internal__/unet.py (hunk 4 above)
Comment (ianstenbit) [reply keras_cv/models/__internal__/unet.py:123]: Does this need to be configured in the final OSS model?

If possible, I'd like to keep config to a minimum unless necessary

File: keras_cv/models/__internal__/unet.py (hunk 3 above)
Comment (ianstenbit) [reply keras_cv/models/__internal__/unet.py:120]: Good call -- removing them throughout.

File: keras_cv/models/__internal__/unet.py (hunk 5 above)
Comment (ianstenbit) [reply keras_cv/models/__internal__/unet.py:20]: done

File: keras_cv/models/__internal__/unet.py [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+    All up and down blocks scale by a factor of two. Skip connections are
+    included.
+
+    All function parameters require curried functions as inputs which return a
+    function that acts on tensors as inputs.
+
+    Args:
+        down_blocks: a list of (filter_count, num_blocks) tuples indicating the
+            number of filters and sub-blocks in each down block
+        up_blocks: a list of filter counts, one for each up block
```
[the remark this answers is not in the document]
Comment (ianstenbit) [reply keras_cv/models/__internal__/unet.py]: done (same for up_blocks)

File: keras_cv/models/__internal__/unet.py (hunk 4 above)
Comment (lengzq) [reply keras_cv/models/__internal__/unet.py:117]: Let's always use synchronized batch norm.

File: keras_cv/models/__internal__/unet.py (hunk 2 above)
Comment (lengzq) [reply keras_cv/models/__internal__/unet.py:109]: Sounds good!

**Review by lengzq — APPROVED** [review]

File: keras_cv/models/__internal__/unet.py (hunk 2 above)
Comment (tanzhenyu) [reply keras_cv/models/__internal__/unet.py:108]: Ah sounds good.

**Review by tanzhenyu — APPROVED** [review]
